vllm/multimodal/processing.py

```diff
@@ -107,86 +115,229 @@ def to_multi_format(data: MultiModalDataDict) -> MultiModalMultiDataDict:
-    # it may be tokenized differently based on whether it is at the start/end
-    # of the string. So, we go through each combination of whether the text
-    # is at the start and end boundaries of the string
+def _max_vocab_token_len(tokenizer: AnyTokenizer) -> int:
+    return max(len(token_text) for token_text in tokenizer.get_vocab())
 
-    # Matches the placeholder when it is in the middle of the string
-    start_id, = encode_no_special_tokens(tokenizer, "a")
-    end_id, = encode_no_special_tokens(tokenizer, "b")
 
-    candidate_basic = encode_no_special_tokens(tokenizer, placeholder_text)
+class _TokenMatch(NamedTuple):
+    start_idx: int
+    end_idx: int
 
-    start_id_, *candidate_a = encode_no_special_tokens(
-        tokenizer,
-        f"a{placeholder_text}",
+
+def find_token_match(
+    token_ids: List[int],
+    match_ids: List[int],
+) -> Optional[_TokenMatch]:
+    """
+    Find the first occurrence of :code:`match_ids` in :code:`token_ids`.
+    """
+    match_len = len(match_ids)
+
+    for start_idx in range(len(token_ids) - match_len + 1):
+        end_idx = start_idx + match_len
+        if token_ids[start_idx:end_idx] == match_ids:
+            return _TokenMatch(start_idx, end_idx)
+
+    return None
+
+
+class _TokenMatchFromTextCandidate(NamedTuple):
+    start_idx: int
+    end_idx: int
+
+    match_text_prefix: str
+    match_text_suffix: str
+
+    @property
+    def distance(self) -> int:
+        return len(self.match_text_prefix) + len(self.match_text_suffix)
+
+
+class _TokenMatchFromText(NamedTuple):
+    start_idx: int
+    end_idx: int
+
+    match_prefix: List[int]
+    match_suffix: List[int]
+
+    match_text_prefix: str
+    match_text_suffix: str
+
+
+def find_token_match_by_text(
+    tokenizer: AnyTokenizer,
+    token_ids: List[int],
+    token_text: str,
+    match_text: str,
+) -> Optional[_TokenMatchFromText]:
+    """
+    Find the first occurrence of the tokenized :code:`match_text` in
+    :code:`token_ids`.
+    """
+    match_ids = _cached_encode(tokenizer, match_text, add_special_tokens=False)
+    if (match := find_token_match(token_ids, match_ids)):
+        return _TokenMatchFromText(
+            match.start_idx,
+            match.end_idx,
+            match_prefix=[],
+            match_suffix=[],
+            match_text_prefix="",
+            match_text_suffix="",
+        )
+
+    # When `match_text` is not mapped to a special token ID,
+    # it may be tokenized differently based on the surrounding tokens
+    # as well as whether it is at the start/end of the string.
+    # Therefore, we need to use `token_text` as a reference.
+    text_start_idx = token_text.find(match_text)
+    if text_start_idx == -1:
+        return None
+
+    text_end_idx = text_start_idx + len(match_text)
+
+    # In case the left/right side of `match_text` is fused with the
+    # string immediately before/after it as a single token
```

Comment:
IMO we should at least make a new method for this case for the code starting from here.